_posts/2019-08-05-image-itk.md

```diff
@@ -0,0 +1,686 @@
+---------------------------
+
+However in this case there are a few challenges to constructing the right Numpy
+-> Numpy function, due to both idiosyncrasies in ITK and Dask Array.  Let's
+look at our function again:
+
+```python
+def richardson_lucy_deconvolution(img, psf, iterations=1):
+    """ Apply deconvolution to a single chunk of data """
+    import itk
+
+    img = img[0, 0, ...]  # remove leading two length-one dimensions
+    psf = psf[0, 0, ...]  # remove leading two length-one dimensions
+
+    image = itk.image_view_from_array(img)   # Convert to ITK object
+    kernel = itk.image_view_from_array(psf)  # Convert to ITK object
+
+    deconvolved = itk.richardson_lucy_deconvolution_image_filter(
+        image,
+        kernel_image=kernel,
+        number_of_iterations=iterations
+    )
+
+    result = itk.array_from_image(deconvolved)  # Convert back to Numpy array
+    result = result[None, None, ...]  # Add back the leading length-one dimensions
+
+    return result
+
+out = da.map_blocks(richardson_lucy_deconvolution, imgs, psf, dtype=np.float32)
+```
+
+This is longer than we would like.
+Instead, we would have preferred to just use the `itk` function directly,
+without all of the steps before and after.
+
+```python
+deconvolved = da.map_blocks(itk.richardson_lucy_deconvolution_image_filter, imgs, psf)
+```
+
+What were the extra steps in our function and why were they necessary?
+
+1.  **Convert to and from ITK Image objects**: ITK functions don't consume and
+    produce Numpy arrays, they consume and produce their own `Image` data
+    structure.  There are convenient functions to convert back and forth,
+    so handling this is straightforward, but it does need to be handled each
+    time.  See [ITK
+    #1136](https://github.com/InsightSoftwareConsortium/ITK/issues/1136) for a
+    feature request that would remove the need for this step.
+
+2.  **Unpack and pack singleton dimensions**: Our Dask arrays have shapes like
+    the following:
+
+    ```
+    Array Shape: (3, 199, 201, 1024, 768)
+    Chunk Shape: (1,   1, 201, 1024, 768)
+    ```
+
+    So our `map_blocks` function gets NumPy arrays of the chunk size,
+    `(1, 1, 201, 1024, 768)`.
+    However, our ITK functions are meant to work on 3d arrays, not 5d arrays,
+    so we need to remove those first two dimensions.
+
+    ```python
+    img = img[0, 0, ...]  # remove leading two length-one dimensions
+    psf = psf[0, 0, ...]  # remove leading two length-one dimensions
+    ```
+
+    And then when we're done, Dask expects to get back 5d arrays like what it
+    provided, so we add these singleton dimensions back in
+
+    ```python
+    result = result[None, None, ...]  # Add back the leading length-one dimensions
+    ```
+
+    Again, this is straightforward for users who are accustomed to NumPy
+    slicing syntax, but does need to be done each time.
+    This adds some friction to our development process,
+    and is another step that can confuse users.
```

Comment:
ITK might be able to support 5-D here (though @thewtex would be able to comment). However the binaries shipped support only a few different ranks of arrays. Also ITK is pretty good about noting this when one tries to make an image with an unsupported rank. This is because at the C++ level the array rank is templated (like a 2-D and a 3-D array are different types). ITK chooses to build out a few, but not all of these array types for various reasons (like to keep the final binary compact). Even if we were able to convert to a higher rank array, we may encounter other issues like the algorithm requires a specific rank. Though this likely has a clear error and would be no different from an error another library might raise. Not sure if or how we should mention this here, just wanted to raise awareness on this point.

Reply:
I think I tried this and things didn't work correctly.  Should I try again @thewtex ?

Reply:
Right I'm not saying the ITK binaries support this now. They were not built with this functionality. I'm saying one could build ITK in a way where this could work. This makes the binaries bigger (hence why it wasn't done according to the error message), but perhaps tolerable if you need it. Just want to make sure we are being fair to ITK here.

Reply:
Just to follow-up here: as @jakirkham mentions, ITK has compile-time optimizations based on the image dimension -- the Python packages are built for 2D and 3D dimensions. But, with the GUFunc support suggested by @mrocklin , these "core dimensions" will naturally be applied to images that have additional array dimensions that may correspond to channels or time.